zerberus/samplepool.cpp

```diff
@@ -0,0 +1,234 @@
+            if (pos > readPos && pos < writePos)
+                  readPos = pos;
+            if (pos >= writePos) {
+                  //qDebug("ERROR: streaming buffer empty! pos %d, writePos %d", pos, writePos);
+                  // TODO Skip reading if already behind
+                  readPosMutex.unlock();
+                  return 0;
+                  }
+            readPosMutex.unlock();
+            return buffer[pos % (STREAM_BUFFER_SIZE * voice->_sample->channel())];
+            }
+      }
+
+void SampleStream::fillBuffer() {
+      if (!streaming)
+            return;
+
+      readPosMutex.lock();
+      unsigned int writePosInBuffer = writePos % (STREAM_BUFFER_SIZE * voice->_sample->channel());
+      if (readPos < backwardSampleCount) {
+            readPosMutex.unlock();
+            return;
+            }
+      unsigned int readBackInBuffer = (readPos - backwardSampleCount) % (STREAM_BUFFER_SIZE * voice->_sample->channel());
+
+      // buffer is full
+      if ((writePos - readPos) >= STREAM_BUFFER_SIZE) {
+            readPosMutex.unlock();
+            return;
+            }
+
+      readPosMutex.unlock();
+
+      sf_count_t toFill;
+
+      if (readBackInBuffer < writePosInBuffer)
+            toFill = (STREAM_BUFFER_SIZE * voice->_sample->channel()) - writePosInBuffer;
+      else
+            toFill = readBackInBuffer - writePosInBuffer;
+
+      toFill /= voice->_sample->channel(); // to fill in frames
+      while (toFill > 0) {
+            // Just to make sure no nasty things happen -> remove when every seems to work good
+            Q_ASSERT(toFill + (writePosInBuffer / (STREAM_BUFFER_SIZE * voice->_sample->channel())) < STREAM_BUFFER_SIZE);
+            unsigned int framesThatShouldBeRead = toFill;
+
+            sf_seek(sf, fileReadPos, SEEK_SET);
+            sf_count_t frames_read;
+
+            if (voice->_looping && (toFill+fileReadPos > voice->_loopEnd)) {
+                  sf_count_t untilLoop =voice->_loopEnd - fileReadPos;
+                  frames_read = sf_readf_short(sf, &buffer[writePosInBuffer], untilLoop);
+                  framesThatShouldBeRead = untilLoop;
+                  }
+            else
+                  frames_read = sf_readf_short(sf, &buffer[writePosInBuffer], toFill);
+
+            writePos += frames_read * voice->_sample->channel();
+            writePosInBuffer = writePos % (STREAM_BUFFER_SIZE * voice->_sample->channel());
+            fileReadPos += frames_read;
+            toFill -= frames_read;
+
+            if (voice->_looping && fileReadPos > voice->_loopEnd)
+                  fileReadPos -= loopDuration;
+
+            if (framesThatShouldBeRead != frames_read)
+                  qDebug("ERROR: reading file with error %s", sf_strerror(sf));
+            }
+      }
+
+
+void BufferThread::run()
+      {
+      while (true) {
+            samplePool->fillSteamBuffers();
+            // TODO use a time based on voices and buffer size!
+            usleep(100);
```

Comment:
Ideally, I guess that this should wait and be woken up when there is actual work to do, meaning filling buffers.